### DISCUSSION TITLE
Last.fm scrobbles don't complete on a stream with crossfade?

### DISCUSSION THREAD

**Discussion by woodruffw** [savonet/liquidsoap#3173] [User Support]
First of all, thank you for creating and maintaining Liquidsoap! It's an incredibly cool tool, not to mention fun to use :slightly_smiling_face: 

I'm opening this because I'm trying to scrobble tracks as they go through my processing pipeline. I'm using `lastfm.submit.full` like so:

```
music = mksafe(crossfade(duration=3.0, playlist(playlist_path, mode="normal")))
music = lastfm.submit.full(user=lastfm_username, password=lastfm_password, music)
```

This results in a "currently playing" scrobble entry as expected, but the "final" scrobble that actually marks the song as fully played never seems to get triggered. Consequently, when the next song begins, the previous "currently playing" entry disappears entirely.

I haven't fully debugged it, but it _appears_ to be caused by the `crossfade` processing phase. When I do this instead:

```
music = mksafe(playlist(playlist_path, mode="normal"))
music = lastfm.submit.full(user=lastfm_username, password=lastfm_password, music)
```

...the scrobbles appear to be persisted correctly. However, it'd be really nice to use it with crossfading!

## Alternatives considered

Using `on_metadata` + `lastfm.submit` also works correctly, even with the crossfade:

```
music.on_metadata(fun(m) -> lastfm.submit(user=lastfm_username, password=lastfm_password, m))
```

However, this only provides the "last played" last.fm record, not the "currently playing indicator." So it's a partial workaround :slightly_smiling_face:

**Comment by toots** [comment]
Hi @woodruffw ! Glad you're enjoying the tool!

Crossfade transitions will remove track marks, for instance when doing crossfades, which will break certain behavior that rely on waiting for a track to be finished.

The easiest solution should be to insert the operator between the `playlist` source and the `crossfade`:

```liquidsoap
music = playlist(playlist_path, mode="normal")
music = lastfm.submit.full(user=lastfm_username, password=lastfm_password, music)
music = mksafe(music)
music = crossfade(duration=3.0, music)
```

**Reply by woodruffw** [reply]
Thanks for the fast reply!

That is indeed an easy solution -- I'll give it a try tonight, thank you!
